### PR TITLE
feat(errors) Split replacer redis keys per dataset

### PR DESCRIPTION
**Pull request by fpacifici**
We will soon have two datasets scheduling replacements concurrently.
This means that the redis keys for the groups to be excluded from queries will have to be disjoint since the replacement may happen at different times.

This just add the dataset name to the redis key (unless this is the events dataset, then instead of migrating the existing key, we add nothing). I don't think there is the need for a migration since the events dataset is being decommissioned.

### CODE REVIEW

**Comment by tkaemming** [comment]
If replacements are being pushed down to the storage level per #778, does this mean that the replacements should be tracked at that level?

What happens if two datasets share the same table?

> I don't think there is the need for a migration since the events dataset is being decommissioned.

Is there a plan here for single tenant and on-premise?

**Comment by fpacifici** [comment]
'does this mean that the replacements should be tracked at that level?'
What do you mean by 'tracked' ?

What I have in mind is a little more complex than putting the replacement in the storage level. 
The storage level is divided into two areas: the storage schema (ReplacingMergeTreeSchema) and the writer (TableWriter, this abstraction is still early on). The schema defines how do deal with data, the writer actually runs DML.

To me in the replacement refactoring, the logic that builds queries, that knows how to apply a DML can be close to the schema, the logic that actually applies writes is in the writer.

There are 1 or more table schemas per dataset, some of those table schema have a writer associated.
If a table is shared between datasets, only one of them defines the writer so all of them can query but only one can write.

**Comment by fpacifici** [comment]
'Is there a plan here for single tenant and on-premise?'

Are you talking about replacement refactoring in general or about the migration from events to errors? In the first case I don't think there is anything to do specifically for on premise. In the second case, yes, at the end of the migration, before we decommission the event table we will provide a way to migrate data

**Comment by tkaemming** [comment]
> What do you mean by 'tracked' ?

Tables (in the ClickHouse sense, regardless of the name of the higher level abstraction) can be referenced by multiple datasets (at least this is my understanding of the discover dataset, but I still need to go back and reread.)

We've also discussed the possibility that a dataset will output the data from it's input source into different output tables (which are still considered part of one "dataset", at least by my interpretation.) For example, subscriptions will have a different access pattern than other queries, and at some point we may need to write data to a different table that shares the same columns as the current table, but has different table options (and is maybe on a different ClickHouse cluster entirely) to improve performance as well as isolating query patterns that are drastically different from one another. Other than the table settings (and hosts that the tables are located on, possibly), everything else is the same — the input data is the same, the way the input data is processed to map to the output format is the same, as far as anyone running a query is concerned the "schema" is the same.

There are two issues with this that I see:

- if a table is shared between two datasets, the replacement state will be associated with one them (which seems to be an essentially arbitrary choice based on which dataset was the one that was blessed as the "writer" at that point, it also implies that a dataset knows about and is dependendent about the existence of other datasets, which doesn't seem like a good thing to me)
- if a dataset contains multiple tables, the replacement state will be associated with all of the tables as a single entity, instead of the ones that have actually had the replacement applied (since replacements will presumably will be applied at different rates to different tables)

I guess my main point is that replacement state is a property of the table itself, regardless of what dataset "owns" it.

> yes, at the end of the migration, before we decommission the event table we will provide a way to migrate data

Yes, this is what I was asking.

**Comment by fpacifici** [comment]
`I guess my main point is that replacement state is a property of the table itself, regardless of what dataset "owns" it.`

Fair, will change replace dataset with table in the key.

About the rest of the comment:
Let's not forget that the existence of the replacer state is very events/errors specific. We are quite far from a general concept of replacer. If the dataset writes on multiple tables, I guess, what the replacer does and what the state will mean will depends a lot on the dataset itself. Maybe it writes on one table and the others are matview, or maybe they are all tables that needs to stay consistent with each other. I think when we want to build a generic replacer we will encounter a whole new type of problems. I am not addressing them now.

About having only one writer per table (whether or not the replacer state is associated with table or dataset), I think allowing multiple writers (thus potentially multiple consumers) to concurrently write to the same table from different datasets is more dangerous than requiring datasets to know each others. Also what we need to avoid is multiple writers writing at the same time, which is something we can enforce without datasets caring about each other.

**Comment by fpacifici** [comment]
Now this is based and depends on https://github.com/getsentry/snuba/pull/781.
As I mentioned there, I should have done that one earlier.
This way setting the state is just a parameter of the replacement processor which is set differently by eerrors and events.  There is no more implicit coupling between read path and write path relying on table names, dataset names, etc.

File: snuba/query/project_extension.py [hunk 1]
```diff
@@ -107,14 +107,22 @@ class ProjectWithGroupsProcessor(ProjectExtensionProcessor):
     2. Taking into consideration groups that should be excluded (groups are excluded because of replacement).
     """
 
+    def __init__(self, project_column: str, replacer_state_name: ReplacerState) -> None:
+        super().__init__(project_column)
+        # This is used to disambiguate datasets in Redis when reading query flags.
+        # If not present, infer the table name from the query.
```
Comment (tkaemming) [inline snuba/query/project_extension.py]: This comment doesn't seem accurate

**Review by tkaemming — APPROVED** [review]

File: snuba/query/project_extension.py (hunk 1 above)
Comment (fpacifici) [reply snuba/query/project_extension.py]: right, it came from the previous version. Will rephrase

**Pull request merged**